### PR TITLE
feat: make org hierarchy depth-agnostic ahead of the territory level

### LINKED ISSUE #47: feat: make org hierarchy depth-agnostic ahead of the "territory" org level
## Context

F3 is adding a sixth level to its org hierarchy. Today an org is one of five kinds — **nation → sector → area → region → AO**. Going forward there will be six, with **territory** between sector and area:

**nation → sector → territory → area → region → AO**

The umbrella work is tracked in [F3-Nation/f3-nation#855](https://github.com/F3-Nation/f3-nation/issues/855). This issue is the `f3-org-map` half of it.

This app reads `/v1/org-chart` and `/v1/org-chart/{id}` from `api.f3nation.com` ([`src/api.ts`](https://github.com/F3-Nation/f3-org-map/blob/main/src/api.ts)). `org_type` is part of that published contract, and it is gaining a value. Because this app deploys independently of the monorepo, **nothing couples our deploy to that migration** — which is why the ordering constraint at the bottom matters.

## Why this is urgent rather than cosmetic

`f3-nation#855` catalogues three places in the monorepo that assume a five-deep tree and return *incomplete* results when it gets deeper. This repo has the same class of bug, but it behaves worse: it returns **confidently wrong** results.

A truncated `LEFT JOIN` yields `NULL`, and null is at least recognizable as missing. Our failure mode instead coerces an unrecognized org type into a **valid-looking one**, so a territory becomes indistinguishable from a real sector or area anywhere downstream. Nothing throws, nothing logs, and the map renders a plausible but incorrect hierarchy.

### The four breakages

| # | Location | What it does today | What happens when `territory` appears |
|---|---|---|---|
| 1 | [`src/orgChart.ts:1`](https://github.com/F3-Nation/f3-org-map/blob/main/src/orgChart.ts#L1) | `export type OrgType = 'nation' \| 'sector' \| 'area' \| 'region' \| 'ao'` | Type no longer describes the API's actual response |
| 2 | [`src/orgChart.ts:38`](https://github.com/F3-Nation/f3-org-map/blob/main/src/orgChart.ts#L38) | `normalizeOrgType` whitelists five strings and returns `fallback` on anything else | `"territory"` is not rejected — it is **silently relabeled** as whatever the caller passed as fallback |
| 3 | [`src/orgChart.ts:86`](https://github.com/F3-Nation/f3-org-map/blob/main/src/orgChart.ts#L86) | `hierarchyTypes: OrgType[] = ['area', 'sector', 'nation']` indexes the API's `hierarchy` array **by position** | The chain is one longer, so every index shifts. Combined with #2, a territory at index 1 is labeled `'sector'`, and the real sector at index 2 is labeled `'nation'` |
| 4 | [`src/state.ts:7`](https://github.com/F3-Nation/f3-org-map/blob/main/src/state.ts#L7) | `levelOrder = ['sector','area','region','ao']` with **numeric** `?level=N` URL state ([`updateUrlState`](https://github.com/F3-Nation/f3-org-map/blob/main/src/state.ts#L11)) | Inserting a tier shifts every index. **Every previously shared or bookmarked org-map link silently resolves to the wrong level** |

Breakages #2 and #3 compound. At [`orgChart.ts:122`](https://github.com/F3-Nation/f3-org-map/blob/main/src/orgChart.ts#L122) the tuple branch reads:

```ts
const parentType = normalizeOrgType(parentTypeRaw, hierarchyTypes[index] ?? 'nation')
```

The API already sends the correct type in the tuple (`[id, name, orgType]`). We take it, fail to recognize `"territory"`, and **discard it in favour of a positional guess that is also now wrong.** The one piece of authoritative data in the payload gets thrown away.

### Downstream consumers of the mislabeled data

Once `buildOrgHierarchy` produces wrong `orgType` values, everything keyed on them is wrong. All in [`src/main.ts`](https://github.com/F3-Nation/f3-org-map/blob/main/src/main.ts):

- [L64-66](https://github.com/F3-Nation/f3-org-map/blob/main/src/main.ts#L64) — three hardcoded layer buttons (`data-level="0|1|2"`, Sectors/Areas/Regions)
- [L388-407](https://github.com/F3-Nation/f3-org-map/blob/main/src/main.ts#L388) — `if (orgType === 'sector') … else if ('area') … else if ('region')` drill-down chain with literal `setCurrentLevelIndex(1|2)` calls
- [L561-563](https://github.com/F3-Nation/f3-org-map/blob/main/src/main.ts#L561) — descendant counts filtered per hardcoded type
- [L657-659](https://github.com/F3-Nation/f3-org-map/blob/main/src/main.ts#L657) — nested ternaries deciding which counts an info panel shows, one branch per level
- [L822-847](https://github.com/F3-Nation/f3-org-map/blob/main/src/main.ts#L822) — `getOrgsForLevel`, including the International-sector special case
- [L946](https://github.com/F3-Nation/f3-org-map/blob/main/src/main.ts#L946) / [L959-960](https://github.com/F3-Nation/f3-org-map/blob/main/src/main.ts#L959) — `currentLevelIndex === 2` magic number and the region-is-a-leaf rule
- [L994](https://github.com/F3-Nation/f3-org-map/blob/main/src/main.ts#L994) — search index restricted to sector/area/region
- [L61](https://github.com/F3-Nation/f3-org-map/blob/main/src/main.ts#L61) — the `Sectors → Areas → Regions → AOs` subtitle

## Proposed approach

Same remedy `f3-nation#855` applies in the monorepo: **one ordered array as the source of truth, everything else derived from it.**

### 1. Derive `OrgType` from an ordered constant

```ts
export const ORG_TYPES = ['ao', 'region', 'area', 'territory', 'sector', 'nation'] as const
export type OrgType = (typeof ORG_TYPES)[number]
```

Order is load-bearing (leaf → root, matching the monorepo's `@acme/shared` array). `levelOrder` in `state.ts` derives from it rather than being written out a second time.

### 2. Make `normalizeOrgType` fail loudly

Change the return type to `OrgType | null` and drop the `fallback` parameter. This is the crux of the whole issue: **an unrecognized org type is data we do not understand, and guessing is what turned a visible failure into an invisible one.** Callers decide explicitly — skip the org, or surface it — but nothing silently invents a type.

### 3. Trust the API's tuple type, delete `hierarchyTypes`

The `hierarchy` entries carry their own `orgType`. Use it. The positional array only ever existed to serve the legacy bare-number form of `hierarchy` (`hiearchy`, note the historical typo, at [`orgChart.ts:29`](https://github.com/F3-Nation/f3-org-map/blob/main/src/orgChart.ts#L29)) — if that form is still in the wild it should skip the entry rather than guess, since a positional guess is now guaranteed wrong at some depth.

### 4. Generate the level UI from the config

Layer buttons, the subtitle, the drill-down chain, the counts, and the info-panel ternaries all become loops or lookups over the ordered array. Replace `currentLevelIndex === 2` and `setCurrentLevelIndex(1|2)` with rank lookups.

### 5. Migrate `?level=N` to a named level

Switch to `?level=area`. Keep reading numeric values for one release, mapped against the **old** four-entry order, so existing bookmarks keep working; then drop it. Without this, inserting territory quietly repoints every link anyone has ever shared.

**Open question for reviewers:** are territories drilled into as their own map layer, or does the drill-down skip them and go sector → area directly? This issue assumes territory is a full layer with its own polygons and buttons. If territories have no meaningful geography of their own, we should decide that now — it changes items 4 and 5.

## Testing

[`src/main.test.ts`](https://github.com/F3-Nation/f3-org-map/blob/main/src/main.test.ts) currently covers `apiGet`, `fuzzyScore`, `convexHull`, and URL-state helpers — `buildOrgHierarchy` and `normalizeOrgType` are untested, which is a fair part of why this went unnoticed.

- Add `buildOrgHierarchy` tests against a **six-level** `/v1/org-chart` fixture, asserting territories keep their own type and ancestors are not shifted
- Add a `normalizeOrgType` test asserting unknown input returns `null` rather than a plausible type
- Extend the URL-state tests to cover a numeric legacy `?level=N` alongside the named form

## Rollout constraint

**This should land before the territory migration runs in `f3-nation`, not after.** The two repos deploy independently, so as soon as a territory row exists in production the deployed map starts mislabeling it — there is no error to page on and no coupling that would hold it back. The fix is behavior-preserving on the current five-level hierarchy, so it can ship early and sit harmlessly.

Also worth noting: F3 keeps growing, so there will be a next level after this one. Done as described, the next one is a one-line change here.

Tracked by F3-Nation/f3-nation#855.

### PR DESCRIPTION
**Pull request by BigGillyStyle**
## Summary

F3's org hierarchy is gaining a sixth level, **territory**, between sector and area (`nation → sector → territory → area → region → AO`). Today this app hardcodes a five-level hierarchy in several places and, worse, silently coerces an unrecognized org type into a *plausible but wrong* one instead of failing loudly — a truncated `hierarchy` positional array would mislabel a territory as a sector or area with nothing thrown and nothing logged.

This PR makes the hierarchy depth-agnostic so it survives territory's insertion (and the next level after that) without further hardcoding:

- `ORG_TYPES` (in `src/orgChart.ts`) is now the single ordered source of truth `OrgType` derives from. `levelOrder`, `layerButtonTypes`, and `drillableTypes` (in `src/state.ts`) are all derived from it instead of duplicated as separate positional arrays.
- `normalizeOrgType` now returns `OrgType | null` instead of silently falling back to a caller-supplied guess. `buildOrgHierarchy` trusts each hierarchy tuple's own `orgType` string and skips (with a `console.warn`) anything it can't recognize — including legacy bare-number `hierarchy`/`hiearchy` entries, which carry no type info and were previously assigned a type purely by array position.
- Layer buttons, the subtitle, the search placeholder, and the info-panel descendant counts in `src/main.ts` are now generated from that config instead of hardcoded per-level branches and a growing ternary chain.
- `?level=` URL state now writes a named level (e.g. `level=region`) and still reads old numeric links, remapped against the pre-territory 4-level order, so existing bookmarks/shared links keep resolving to the correct level.
- Fixed a latent out-of-bounds crash: deep-linking directly to an `ao` org with no `level` param would compute an index one past the end of `levelOrder`.

## Test plan

- [x] `npx tsc --noEmit` — clean
- [x] `npm run lint` (Biome) — clean
- [x] `npm run build` — clean
- [x] `npx vitest run` — 17/17 passing, including new `src/orgChart.test.ts` coverage (six-level hierarchy fixture, `normalizeOrgType` null-handling, legacy hierarchy skipping) and updated `src/main.test.ts` URL-state coverage (named levels, legacy numeric levels, AO-clamp regression)
- [x] Verified in-browser against the real production API (`api.f3nation.com`, 551 live orgs): Territories button/subtitle/search-placeholder render correctly, drill-down correctly advances sector → territory (not the old off-by-one into area), info-panel counts are correct at each level, and both named and legacy-numeric `?level=` deep links resolve to the correct level

Fixes #47.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added territory-level support across organization maps and hierarchy navigation.
  * Added layer switching, territory-aware breadcrumbs, and fuzzy search.
  * Organization information now includes territory counts.
  * Deep links support named hierarchy levels while retaining compatibility with legacy numeric links.
  * Improved handling of invalid organization data and default navigation levels.

* **Documentation**
  * Updated feature documentation to cover territory polygons, navigation, layer switching, and fuzzy search.

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -403,27 +417,14 @@ function navigateToOrg(org: Org) {
...
-    )
-    selectedPath.length = 0
-    selectedPath.push(...path)
-    setCurrentLevelIndex(levelIndex)
+  selectedPath.length = 0
+  selectedPath.push(...buildBreadcrumbPath(org, getOrgPath(org)))
+
+  if (drillableTypes.includes(org.orgType)) {
+    // Drill into the next level: breadcrumb includes the selected org itself.
+    setCurrentLevelIndex(levelIndex + 1)
```
Comment (BigGillyStyle) [inline src/main.ts]: **🔴 Critical** · code-reviewer

`navigateToOrg` now unconditionally advances to `levelIndex + 1` for any drillable org type, and `getCurrentLevelOrgs` (line 899) filters strictly on `org.orgType === level && org.parentId === parent.id` with no fallback when that level is empty.

If a sector's direct children are still `area` orgs (no `territory` records yet — plausible during a gradual backend rollout of the new level, or for any sector left un-migrated), clicking that sector renders `No territories available.` and the entire subtree below it (areas, regions) becomes unreachable by click. This regresses the primary navigation path for any not-yet-migrated sector, whereas before this PR the same click correctly landed on `area`.

**Suggested fix:** instead of advancing to a fixed `levelIndex + 1`, advance to the next level in `levelOrder` that actually has children under `parent` (skip levels with zero matches), or explicitly special-case the transitional period where `territory` may not yet be populated for every sector.

— _Claude Code_

File: src/orgChart.ts [hunk 2]
```diff
@@ -120,26 +120,36 @@ export function buildOrgHierarchy(items: OrgChartItem[]): Org[] {
     const itemId = item.id ?? item.orgId
     if (itemId == null) return
     const parentId = getOrgParentId(item)
```
Comment (BigGillyStyle) [inline src/orgChart.ts]: **🟠 Important** · code-reviewer, pr-test-analyzer

`getOrgParentId(item)` reads `hierarchy[0][0]` unconditionally as the child's `parentId`, before the loop below has a chance to validate that ancestor's `orgType`. When that ancestor turns out to be unrecognized (or is a legacy bare-number entry) and gets skipped, the child org still ends up with a `parentId` pointing at an org that was never created in `orgMap`.

This is actually demonstrated by the PR's own new test (`src/orgChart.test.ts`, "skips a hierarchy-tuple ancestor with an unrecognized type"): it asserts `byId.has(20)` is `false`, but never checks that org 30's `parentId` is still `20` — a dangling reference. Downstream, `getOrgPath` (`src/main.ts:389`) walks `parentId` via `orgById.get(...)`, hits the missing id, and silently truncates the ancestor chain — breaking breadcrumbs and `getCurrentLevelOrgs`'s parent-based filtering for that node, with only a `console.warn` from the *ancestor* skip and no indication the *child* is now orphaned.

**Suggested fix:** when a hierarchy-tuple ancestor is skipped, either null out the affected child's `parentId` (so it's treated as a root rather than silently orphaned deeper in the tree), or defer setting the child's `parentId` until the referenced ancestor is confirmed valid.

— _Claude Code_

File: src/main.ts [hunk 3]
```diff
@@ -870,11 +875,16 @@ function getCurrentLevelOrgs(): Org[] {
...
+  // descendants (not just direct children), since International's structure
+  // doesn't nest cleanly through area.
+  if (
+    isSectorInternational(parent) &&
+    (level === 'territory' || level === 'area' || level === 'region')
+  ) {
     const internationalDescendants = getDescendantOrgIds(parent.id)
     return [...orgById.values()].filter(
-      (org) => org.orgType === 'region' && internationalDescendants.includes(org.id),
+      (org) => org.orgType === level && internationalDescendants.includes(org.id),
```
Comment (BigGillyStyle) [inline src/main.ts]: **🟠 Important** · code-reviewer

_(Anchored here because the actual defect, `renderBreadcrumb`'s click handler at `src/main.ts:930-931`, isn't touched by this diff and GitHub won't accept a comment there — but it's a direct consequence of the change in this hunk.)_

`renderBreadcrumb`'s click handler does:
```js
selectedPath.length = depth + 1
setCurrentLevelIndex(depth + 1)
```
This assumes `selectedPath[depth].orgType === levelOrder[depth]` — that the breadcrumb array and the level-order array stay positionally aligned. That held when the hierarchy was a tight, fixed chain, but `buildBreadcrumbPath`/`getOrgPath` build `selectedPath` from the org's *actual* ancestor chain, which can now have gaps relative to `levelOrder` — including right here, via International's territory/area/region special-casing, whose structure the comment above notes "doesn't nest cleanly through area."

**Concrete failure:** `selectedPath = [sector, area]` (no territory in between, e.g. a not-yet-migrated sector, or an International descendant reached without an intermediate territory). Clicking the `area` crumb (depth `1`) sets `currentLevelIndex = 2`, which is `levelOrder[2] === 'area'` — but `getCurrentLevelOrgs` then uses `parent = area` and filters `orgType === 'area' && parentId === area.id`, which returns nothing (an area isn't its own child).

**Suggested fix:** derive the index from the clicked org's actual type instead of its position: `setCurrentLevelIndex(levelOrder.indexOf(selectedPath[depth].orgType) + 1)`.

— _Claude Code_

File: src/main.ts [hunk 4]
```diff
@@ -803,7 +805,10 @@ async function loadOrgInfo(org: Org) {
     const existing = orgById.get(org.id)
     if (existing) {
       existing.name = data.name ?? existing.name
-      existing.orgType = normalizeOrgType(data.orgType, existing.orgType)
+      const refreshedType = normalizeOrgType(data.orgType)
+      if (refreshedType != null) {
+        existing.orgType = refreshedType
+      }
```
Comment (BigGillyStyle) [inline src/main.ts]: **🟠 Important** · silent-failure-hunter

Every other `normalizeOrgType` call site added by this PR pairs a `null` result with a `console.warn` naming the org and the rejected raw value (`src/orgChart.ts:127, 144-147`). This call site drops the `else` branch entirely: when the org-detail API returns an `orgType` string `normalizeOrgType` can't recognize, `existing.orgType` is silently left at its previous value with no log at all.

**Concrete scenario:** the backend renames/adds an org type (exactly what's happening with `territory`) and the detail endpoint for some org starts returning a string this build doesn't yet recognize. The map keeps showing/filtering that org under its old type indefinitely, and there's no console signal to tell a developer the frontend and API have drifted — silently defeating the PR's own stated goal of never guessing/silently coercing an unrecognized type.

**Suggested fix:**
```suggestion
      const refreshedType = normalizeOrgType(data.orgType)
      if (refreshedType != null) {
        existing.orgType = refreshedType
      } else {
        console.warn(`Ignoring unrecognized orgType for org ${org.id}:`, data.orgType)
      }
```

— _Claude Code_

File: src/orgChart.ts [hunk 5]
```diff
@@ -1,4 +1,14 @@
-export type OrgType = 'nation' | 'sector' | 'area' | 'region' | 'ao'
+export const ORG_TYPES = ['ao', 'region', 'area', 'territory', 'sector', 'nation'] as const
```
Comment (BigGillyStyle) [inline src/orgChart.ts]: **🟡 Suggestion** · type-design-analyzer, comment-analyzer

`ORG_TYPES`'s ordering (leaf-to-root: `ao` first, `nation` last) is load-bearing — `levelOrder` in `src/state.ts` derives root-to-leaf navigation order by reversing it, and the PR's whole premise is that inserting a future level here is the only change needed. But that convention is only documented at the *consumption* site (`state.ts`'s comment on `levelOrder`), not here at the source of truth. A future editor adding a level to the wrong end of this array gets no compile-time signal and no local comment warning them — just a silently wrong `levelOrder`.

**Suggested fix:**
```suggestion
// Leaf-to-root order (ao → nation). src/state.ts reverses this to derive
// root-to-leaf navigation order — insert a new level here in the position
// matching its place in the real hierarchy.
export const ORG_TYPES = ['ao', 'region', 'area', 'territory', 'sector', 'nation'] as const
```

— _Claude Code_

File: src/state.ts [hunk 6]
```diff
@@ -2,17 +2,53 @@ export function setCurrentLevelIndex(val: number) {
...
+// truth so inserting a new level there is the only change this file needs.
+export const levelOrder: OrgType[] = [...ORG_TYPES].filter((type) => type !== 'nation').reverse()
+
+// Levels that get a layer button, are searchable, and show an info-panel
+// count. 'ao' is deliberately excluded: it has no button/search entry today.
+export const layerButtonTypes: OrgType[] = levelOrder.filter((type) => type !== 'ao')
+
+// Levels where selecting an org advances the drill-down to the next level.
+// The last layer-button level (region) is a view-only leaf: it stays at its
+// own level and shows siblings instead.
```
Comment (BigGillyStyle) [inline src/state.ts]: **🟡 Suggestion** · comment-analyzer

This comment names `region` as *the* view-only leaf, but the actual invariant `drillableTypes = layerButtonTypes.slice(0, -1)` encodes is positional: whichever type ends up last in `layerButtonTypes` is treated as the view-only leaf. Today that happens to be `region`, but the comment will silently misdescribe the code the next time a level is inserted after region (e.g. between region and `ao`) — a reader would trust the comment over the slice and be wrong.

**Suggested fix:**
```suggestion
// Levels where selecting an org advances the drill-down to the next level.
// Whichever type is last in `layerButtonTypes` (currently region) is a
// view-only leaf: it stays at its own level and shows siblings instead.
```

— _Claude Code_

File: src/main.ts [hunk 7]
```diff
@@ -575,17 +576,20 @@ function renderInfo(org: Org, detail?: OrgInfo) {
     aoCount += metrics.aos
     locationsCount += metrics.locations
   })
-  const sectorCount = descendantOrgs.filter((item) => item.orgType === 'sector').length
-  const areaCount = descendantOrgs.filter((item) => item.orgType === 'area').length
-  const regionCount = descendantOrgs.filter((item) => item.orgType === 'region').length
-  const formattedAreaCount = formatNumber(areaCount)
-  const formattedRegionCount = formatNumber(regionCount)
-  const formattedSectorCount = formatNumber(sectorCount)
+  const currentType = detail?.orgType ?? org.orgType
```
[the remark this answers is not in the document]
Comment (BigGillyStyle) [reply src/main.ts]: Confirmed — `detail?.orgType` was the raw cached API value, not the validated one used elsewhere. Fixed by routing it through `normalizeOrgType` (5547d1b).

— _Claude Code_

File: src/main.ts (hunk 1 above)
Comment (BigGillyStyle) [reply src/main.ts]: Confirmed — this breaks navigation for any sector without territory records yet, which is the norm until the backend rollout completes. Fixed by having navigation search forward for the next level that actually has children (`nextLevelIndexWithOrgs`), reusing the same lookup `getCurrentLevelOrgs` uses (including the International-descendants special case) via a new shared `getOrgsAtLevel` helper (5547d1b).

— _Claude Code_

File: src/orgChart.ts (hunk 2 above)
Comment (BigGillyStyle) [reply src/orgChart.ts]: Confirmed dangling reference — and good catch on the existing test not asserting it. Fixed `getOrgParentId` to fall back to `null` (root) when the immediate hierarchy ancestor's type isn't recognized, instead of pointing at an org that's never created. Strengthened the existing test to assert `byId.get(30).parentId` is `null` (5547d1b).

— _Claude Code_

File: src/main.ts (hunk 3 above)
Comment (BigGillyStyle) [reply src/main.ts]: Confirmed — fixed by deriving the level index from the clicked breadcrumb org's actual type (`levelOrder.indexOf(org.orgType) + 1`) rather than its position in `selectedPath`, per your suggested fix (5547d1b).

— _Claude Code_

File: src/main.ts (hunk 4 above)
Comment (BigGillyStyle) [reply src/main.ts]: Confirmed — added the missing `console.warn` to match the pattern used everywhere else `normalizeOrgType` can reject a value (5547d1b).

— _Claude Code_

File: src/orgChart.ts (hunk 5 above)
Comment (BigGillyStyle) [reply src/orgChart.ts]: Good catch — added the ordering note at the source of truth, exactly as suggested (5547d1b).

— _Claude Code_

File: src/state.ts (hunk 6 above)
Comment (BigGillyStyle) [reply src/state.ts]: Agreed, reworded to describe the positional invariant instead of naming region specifically (5547d1b).

— _Claude Code_

**Comment by taterhead247** [comment]
@BigGillyStyle dis you see I put a PR in the mono repo to move this app there and include this functionality in it?

**Comment by BigGillyStyle** [comment]
> @BigGillyStyle dis you see I put a PR in the mono repo to move this app there and include this functionality in it?

I submitted this before I realized what your PR was doing. I'll just hold this here and wait until your PR merges. I imagine Claude can transfer (or just re-do) this code change there.

**Comment by taterhead247** [comment]
This work is completed on https://github.com/F3-Nation/f3-nation
